### ISSUE TITLE
How to get bounding boxes from YOLOv8 instance object detection output in TFLite interpretor?

### ISSUE THREAD

**Issue by noahweber1** [ultralytics/ultralytics#5192]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, this is somewhat similiar to https://github.com/ultralytics/ultralytics/issues/5050

but I am asking concretely for bounding boxes (x,y, heighth, width)

I have converted a YOLO8 object detection model to a quantized edge TPU using https://docs.ultralytics.com/modes/export/

The resulting model, when trying to predict on the same input images of size (480,640,3), gives these two sensors:

common.output_tensor(interpreter, 0)[0].shape
(41, 8400)
common.output_tensor(interpreter, 1)[0].shape
(160, 160, 32)

I suspect that the first tensor is the flattened outputs and the second one is the feature map.

My question is, how to unflatten this?

**Comment by glenn-jocher** [comment]
@noahweber1 hello, thank you for raising this question. The first tensor you have obtained, with shape (41, 8400), is the prediction output. The second tensor, with shape (160, 160, 32), is indeed an intermediate feature map.

The output tensor contains multiple potential bounding boxes. Each position in the output tensor corresponds to a logical grid position in the input image, and each position can predict multiple bounding boxes. 

For YOLOv8, each predicted bounding box representation consists of multiple components: the (x,y) coordinates of the center of the bounding box, the width and height of the bounding box, the objectness score, and the class scores. These are flattened into a single dimension in the output tensor.

To unflatten these predictions, you would reshape this tensor back into the shape of the logical grid used by YOLOv8 and the number of bounding box predictions per grid point. You would also split the objectness scores and class scores into separate dimensions. 

After reshaping and splitting, you would apply the Sigmoid function to the (x, y) coordinates and objectness score, and the Softmax function to the class scores to convert them into probabilities. 

Finally, the width and height of the bounding boxes would be transformed by an exponential function, as they are predicted on a log scale by the model. 

In the end, the unflattening process will give you the bounding boxes you mentioned: (x,y,width,height), along with the objectness score and class probabilities, which are essential for object detection tasks.

Please note that this explanation assumes a basic understanding of the working principles and output formatting of YOLOv8 models. 

Feel free to reach out for further clarification.

**Comment by noahweber1** [comment]
Thank you @glenn-jocher 

Based on your explanation the following is the solution, could you confirm that this is the case?:

(I would then extract the n_classes, n_anchors and grid size from the YOLO8 model that was used.

```

# Constants these should be extracted from the configuration
num_classes = 5  # Number of object classes
num_anchors = 3  # Number of anchor boxes
grid_size = 40  # Size of the output grid (e.g., if your output is a 40x40 grid)


# Reshape the tensor
# The shape would be [grid_size, grid_size, num_anchors, 4+1+num_classes]
reshaped_tensor = output_tensor.view(grid_size, grid_size, num_anchors, 4 + 1 + num_classes)

# Split the tensor into its components
box_centers = reshaped_tensor[..., :2]  # x, y
box_sizes = reshaped_tensor[..., 2:4]  # width, height
objectness = reshaped_tensor[..., 4:5]  # objectness score
class_scores = reshaped_tensor[..., 5:]  # class scores

# Apply the activation functions
box_centers = torch.sigmoid(box_centers)  # Sigmoid for x, y
objectness = torch.sigmoid(objectness)  # Sigmoid for objectness
class_scores = F.softmax(class_scores, dim=-1)  # Softmax for class scores

# Transform box sizes
box_sizes = torch.exp(box_sizes)  # exponential for width, height
```

**Comment by noahweber1** [comment]
thanks again @glenn-jocher just a final question.

Yolo8 is supposed to be anchor free algorithm. However looking at the configuration files or directly in the model architecture this parameter is not available exposed.

I looked for default and here it claims its 3 https://pub.towardsai.net/understanding-hyper-parameter-tuning-of-yolos-82aec5f6e7b3

My question is: where to look exactly for num_anchors.

**Comment by glenn-jocher** [comment]
Hello @noahweber1,

Thank you for your question. It's indeed true that YOLOv8 uses an anchor-free approach contrary to its previous versions. This means it doesn't rely on preset anchor boxes to make predictions. Instead, it uses a sliding window approach on various scales to identify objects of different sizes. Therefore, the concept of 'num_anchors' doesn't apply here.

The parameter is still present in the configuration for backward compatibility with previous versions and to maintain a unified interface. If you see a default value of 3 (or any other value), it doesn't affect the predictions made by YOLOv8.

The article reference you provided might contain some confusions, as it describes hyperparameters for previous versions of YOLO, which used anchor-based predictions. 

In summary, with YOLOv8 you don't need to worry about the 'num_anchors' parameter as the model automatically detects objects of varying sizes without relying on predefined anchor boxes. I hope this clears up your query. If you have further questions, feel free to ask!
